leader-election-impl/src/main/java/com/palantir/leader/lease/LeadershipLeaseLeaderElectionService.java

```diff
@@ -0,0 +1,102 @@
+ */
+
+package com.palantir.leader.lease;
+
+import java.time.Duration;
+import java.util.Optional;
+import java.util.Set;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.net.HostAndPort;
+import com.palantir.leader.LeaderElectionService;
+import com.palantir.leader.PingableLeader;
+
+/**
+ * A leadership election service which implements 'leadership leases' on top of an underlying leader election service.
+ * <p>
+ * Naively, one must verify their leadership before every call against a quorum of nodes. This works, but adds constant
+ * extra RPC load. One can improve this status quo by caching leadership. Nodes take a time X as their 'expiry' time
+ * and a time 'Y' << 'X' as their refresh time. After becoming leader in the underlying service, one must wait for
+ * at least 'X' time before returning their result. Now, once leader, one may check that they are still the leader
+ * by polling every 'Y' time rather than every single check.
+ * <p>
+ * While this does rely on wall clock time, it uses the high precision now counters rather than the usual mutable
+ * wall clock time. This means that breaking the leadership lease would require (on default settings) one computer's
+ * clock to run twice as fast as another's.
+ */
+public final class LeadershipLeaseLeaderElectionService implements LeaderElectionService {
+    private final LeaderElectionService delegate;
+    private final LoadingCache<LeadershipToken, LeadershipLeaseLeadershipToken> leaseTokens = Caffeine.newBuilder()
+            .maximumSize(1)
+            .build(this::createLeaseToken);
+
+    private final Duration leaseRefresh;
+    private final Duration leaseExpiry;
+
+    @VisibleForTesting
+    LeadershipLeaseLeaderElectionService(
+            LeaderElectionService delegate,
+            Duration leaseRefresh,
+            Duration leaseExpiry) {
+        this.delegate = delegate;
+        this.leaseRefresh = leaseRefresh;
+        this.leaseExpiry = leaseExpiry;
+    }
+
+    // this method exists in order to provide a rolling migration mechanism for migrating to schema leases
+    public static LeaderElectionService wrapWithoutLeasing(LeaderElectionService delegate) {
+        return new LeadershipLeaseLeaderElectionService(delegate, Duration.ZERO, Duration.ofSeconds(2));
+    }
+
+    public static LeaderElectionService wrap(LeaderElectionService delegate) {
+        return new LeadershipLeaseLeaderElectionService(delegate, Duration.ofSeconds(1), Duration.ofSeconds(2));
+    }
+
+    @Override
+    public LeadershipToken blockOnBecomingLeader() throws InterruptedException {
+        return leaseTokens.get(delegate.blockOnBecomingLeader());
+    }
+
+    @Override
+    public Optional<LeadershipToken> getCurrentTokenIfLeading() {
+        return delegate.getCurrentTokenIfLeading().map(leaseTokens::get);
```

Comment:
it feels a bit goofy for this to be blocking in the way it is. Should probably have the cache contain an Entry class which only becomes valid past a point (+ tests).

Reply:
fixed by just being like getIfPresent

Reply:
If this method returns Optional.empty, subsequent calls will always return Optional.empty unless `blockOnBecomingLeader()` is called in between. I have checked the usage in `AwaitingLeadershipProxy` and the contract stated in the `LeaderElectionService` interface, and that is inline with this implementation.